### ISSUE TITLE
TryAddEnumerable for HttpClients

### ISSUE THREAD

**Issue by ArminShoeibi** [dotnet/runtime#80202]
Hi there,

We have 5+ TypedHttpClients that implement an identical interface, Is there a way to register them like regular services?
Because I didn't find any related method, is it possible if there is no way?

**Comment by ArminShoeibi** [comment]
@karelz

**Comment by ArminShoeibi** [comment]
Desired:
```
interface OrderInfoHttpClient 
{
    TicketInfo GetTicketInfo();
}

class DomesticFlightHttpClient : OrderInfoHttpClient
{
    public TicketInfo GetTicketInfo()
    {
    }
}

class HotelHttpClient : OrderInfoHttpClient
{
    public TicketInfo GetTicketInfo()
    {
    }
}

public class OrderInfoService
{
    private readonly IEnumerable<OrderInfoHttpClient> _i1HttpClients;

    public OrderInfoService(IEnumerable<OrderInfoHttpClient> i1HttpClients)
    {
        _i1HttpClients = i1HttpClients;

        foreach (var item in _i1HttpClients)
        {
            //.....
        }
    }
}
```

**Comment by teo-tsirpanis** [comment]
c.c. @dotnet/area-extensions-dependencyinjection

**Comment by karelz** [comment]
@CarnaViire can you please comment? I am afraid I am not sure I understand what's being asked here ...

**Comment by ArminShoeibi** [comment]
> @CarnaViire can you please comment? I am afraid I am not sure I understand what's being asked here ...

Hey Karel, assume this code, I want to register many classes as TypedHttpClient but with one identical interface.

```
builder.Services.AddHttpClient<IOrderInfoHttpClient, DomesticFlightHttpClient>(hc =>
{
    hc.BaseAddress = new Uri("");
});

builder.Services.AddHttpClient<IOrderInfoHttpClient, HotelHttpClient>(hc =>
{
    hc.BaseAddress = new Uri("");
});

builder.Services.AddHttpClient<IOrderInfoHttpClient, TrainHttpClient>(hc =>
{
    hc.BaseAddress = new Uri("");
});
```

and after registration I want to inject `IEnumerable<IOrderInfoHttpClient>` to constructors.

**Comment by CarnaViire** [comment]
@ArminShoeibi The short answer to your original question is no, it's currently not possible with TypedClient registration API, as only one typed client implementation can be registered per interface.

I'd like to understand your use case better to suggest a reasonable workaround.
Do you require any additional configuration in your client registrations beside setting BaseAddress?
I wonder whether you can reuse `System.Net.Http.HttpClient` between your services (DomesticFlightHttpClient etc).

**Comment by ArminShoeibi** [comment]
> 

Thanks Natalia,
No, we just set Timeout and BaseAddress for each of them, and we can set a constant value to Timeout for all of them.

**Comment by CarnaViire** [comment]
TL;DR: you *can* actually achieve what you need with current APIs, you just need to specify a distinct name per implementation, when registering your typed client:

```
services.AddHttpClient<ITypedClient, ClientA>("A", client => Console.WriteLine($"Configuring HttpClient#{client.GetHashCode()} for ClientA"));
services.AddHttpClient<ITypedClient, ClientB>("B", client => Console.WriteLine($"Configuring HttpClient#{client.GetHashCode()} for ClientB"));

...

var typedClients = provider.GetServices<ITypedClient>();
foreach (var client in typedClients)
{
    client.Foo();
}
```

Explanation:

Turns out I was not completely right when I said "only one typed client implementation can be registered per interface", because I forgot about this overload: [AddHttpClient<TClient,TImplementation>(this IServiceCollection services, String name, Action<HttpClient> configureClient)](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.httpclientfactoryservicecollectionextensions.addhttpclient?view=dotnet-plat-ext-7.0#microsoft-extensions-dependencyinjection-httpclientfactoryservicecollectionextensions-addhttpclient-2(microsoft-extensions-dependencyinjection-iservicecollection-system-string-system-action((system-net-http-httpclient)))). What actually happens is that, by default, name of the interface type is used as a name (key) for configuration. If you specify the name manually in the overload above, this name will be used as a key instead, which will allow you to have distinct configurations for each of your Typed client implementations.

You can even write some extension method in your code that would use e.g. name of the implementation type as a configuration name.

Let me know if this helps @ArminShoeibi

**Comment by ArminShoeibi** [comment]
> You can even write some extension method in your code that would use e.g. name of the implementation type as a configuration name.
> 
> Let me know if this helps @ArminShoeibi

Thanks, I'll test it and inform you.

**Comment by ArminShoeibi** [comment]
> Let me know if this helps

It works like a charm, Thank you @CarnaViire
